# Day 2 - Linear Regression

## Introduction

Linear Regression is a supervised learning algorithm used to predict continuous numerical values.

It learns the relationship between input features and a numeric target by fitting the best possible linear relationship between them.

## Objective

The goal of this notebook is to build a Linear Regression model, make predictions, evaluate its performance using regression metrics, and interpret the model results.

## What is Linear Regression?

Linear Regression is one of the simplest supervised machine learning algorithms.

It predicts continuous numeric values by learning the relationship between the input features and the target variable.

## Dataset

I used the Student Performance dataset.

The target variable is total_score, which is a continuous numeric value.

Therefore, this is a regression problem.

In [1]:
import pandas as pd 
import numpy as np 
from sklearn.model_selection import train_test_split 
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

In [3]:
df = pd.read_csv("student_performance.csv")

df.head()

,student_id,weekly_self_study_hours,attendance_percentage,class_participation,total_score,grade
0,1,18.5,95.6,3.8,97.9,A
1,2,14.0,80.0,2.5,83.9,B
2,3,19.5,86.3,5.3,100.0,A
3,4,25.7,70.2,7.0,100.0,A
4,5,13.4,81.9,6.9,92.0,A


## Dataset Information

I checked the dataset shape, column names, data types, and missing values before building the model.

In [4]:
df.shape

(1000000, 6)

In [5]:
df.columns

Index(['student_id', 'weekly_self_study_hours', 'attendance_percentage',
       'class_participation', 'total_score', 'grade'],
      dtype='str')

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1000000 entries, 0 to 999999
Data columns (total 6 columns):
 #   Column                   Non-Null Count    Dtype  
---  ------                   --------------    -----  
 0   student_id               1000000 non-null  int64  
 1   weekly_self_study_hours  1000000 non-null  float64
 2   attendance_percentage    1000000 non-null  float64
 3   class_participation      1000000 non-null  float64
 4   total_score              1000000 non-null  float64
 5   grade                    1000000 non-null  str    
dtypes: float64(4), int64(1), str(1)
memory usage: 45.8 MB


In [9]:
df.isnull().sum()

student_id                 0
weekly_self_study_hours    0
attendance_percentage      0
class_participation        0
total_score                0
grade                      0
dtype: int64

### Observation

The dataset contains five numerical columns and one categorical column.

There are no missing values in the dataset.

## Dataset

I used the Student Performance dataset.

The target variable is **total_score**, which is a continuous numeric value.



# Hands-On Lab: Predicting a Continuous Value

## Step 1: Select Features and Target

The target variable is **total_score**.

The selected features are:

- weekly_self_study_hours
- attendance_percentage
- class_participation

These features are expected to influence the student's total score.

In [13]:
X =df[[
    "weekly_self_study_hours",
    "attendance_percentage",
    "class_participation"
]]
y = df["total_score"]

x contain the selected input features
y contain  the total_score which is numeric target variable


## Train and Test Split 
The dataset is divided into training and testing sets
80% of the data is used for training and 20% is reserved for testing 

In [17]:
x_train,x_test,y_train,y_test = train_test_split(
    X,
    y,
    test_size = 0.2
, random_state = 42
)

In [18]:
print("x train shape",x_train.shape)
print("x test shape",x_test.shape)
print("y train shape",y_train.shape)
print("y test shape",y_test.shape)

x train shape (800000, 3)
x test shape (200000, 3)
y train shape (800000,)
y test shape (200000,)


In [21]:
print(len(x_train)==len(y_train))
print(len(x_test)==len(y_test))

True
True


## Observation
The dataset was successfully divided into training and testing sets 
The number of rows in x_train = number of rows in y_train 
and the number of rows in x_test = number of rows in y_test

## Step 1 : Train The Linear Regression Model 
A Linear Regression model is created and trained using the training dataset.

In [35]:
Model = LinearRegression()
Model.fit(x_train, y_train)

print("Model trained successfully")

Model trained successfully


### Observation

The Linear Regression model was successfully trained using the training dataset.

After training, the model calculated one coefficient for each input feature and one intercept value. These values represent the mathematical relationship between the selected features and the target variable.

The output shows that the model was trained using three features:
- weekly_self_study_hours
- attendance_percentage
- class_participation

The model also learned an intercept value (56.95), which represents the predicted total score when all feature values are zero.

The coefficients indicate how much the predicted total score changes when a feature increases by one unit while keeping the other features constant.

The trained model is now ready to make predictions on unseen test data.

## Step 2: Report the Model Coefficients

After training the model, I examined the coefficients and the intercept to understand how each feature affects the prediction.

In [24]:
print("Coefficients:")
print(Model.coef_)

print()

print("Intercept:")
print(Model.intercept_)

Coefficients:
[ 1.81682003e+00  4.01750743e-04 -2.75562094e-03]

Intercept:
56.953557444770254


In [38]:
coefficients = pd.DataFrame({
    "Feature": X.columns,
    "Coefficient": Model.coef_
})

coefficients["Absolute"] = coefficients["Coefficient"].abs()

coefficients.sort_values("Absolute", ascending=False)

,Feature,Coefficient,Absolute
0,weekly_self_study_hours,1.816820,1.816820
2,class_participation,-0.002756,0.002756
1,attendance_percentage,0.000402,0.000402


### Observation

The feature with the strongest effect is `weekly_self_study_hours` because it has the largest absolute coefficient.

Its coefficient is approximately 1.82, which means that increasing weekly self-study time by one hour increases the predicted total score by about 1.82 points, while keeping the other features constant.

Attendance percentage has a very small positive effect, while class participation has a very small negative effect in this model.

## Step 3: Generate Predictions

The trained model is used to predict the total scores of the students in the test dataset.

In [25]:
PRE = Model.predict(x_test)
PRE[:10]

array([64.05954219, 88.03598894, 72.79127762, 84.24282459, 58.60876743,
       63.88168567, 64.42332547, 64.60985945, 85.14087917, 79.67978296])

In [27]:
COMP = pd.DataFrame({
"actual":y_test.values ,
"predicted":PRE

})
COMP.head(10)

,actual,predicted
0,53.9,64.059542
1,77.3,88.035989
2,75.0,72.791278
3,85.7,84.242825
4,30.8,58.608767
5,62.3,63.881686
6,74.0,64.423325
7,68.9,64.609859
8,83.7,85.140879
9,95.5,79.679783


### Observation

The model generated predictions for the test data.

Some predicted values are close to the actual scores, while others show larger differences. The evaluation metrics are used to measure the model's overall performance.

## Step 3: Model Evaluation

The model performance is evaluated using three regression metrics:

- Mean Absolute Error (MAE)
- Root Mean Squared Error (RMSE)
- R² Score

In [36]:
PRE = Model.predict(x_test)

In [37]:
mae = mean_absolute_error(y_test,PRE)

rmse = np.sqrt(mean_squared_error(y_test,PRE))

r2 = r2_score(y_test,PRE)

print("MAE:", mae)
print("RMSE:", rmse)
print("R²:", r2)

MAE: 7.161333476729989
RMSE: 8.996400873369835
R²: 0.6600425865640975


## Step 4: Compare with a Baseline

In [30]:
baseline = np.full(len(y_test), y_train.mean())

baseline_rmse = np.sqrt(
    mean_squared_error(y_test, baseline)
)

print("Baseline RMSE:", baseline_rmse)
print("Model RMSE:", rmse)

Baseline RMSE: 15.429723444712895
Model RMSE: 8.996400873369835


### Observation

The baseline model achieved an RMSE of **15.43**, while the Linear Regression model achieved an RMSE of **9.00**.

Since the model's RMSE is significantly lower than the baseline RMSE, the Linear Regression model provides more accurate predictions.

This shows that the model has successfully learned useful patterns from the training data instead of simply predicting the average score for every student.

## Step 5: Interpretation of Results

The Linear Regression model performed well on the student performance dataset.

The model achieved an **MAE of 7.16**, meaning that the predicted scores differ from the actual scores by about **7 points on average**.

The **RMSE of 9.00** indicates that the overall prediction error is relatively low, while the **R² score of 0.66** shows that the model explains about **66% of the variation** in students' total scores.

Compared with the baseline model (RMSE = **15.43**), the Linear Regression model achieved a much lower prediction error. This confirms that the model learned meaningful relationships between the input features and the target variable and provides better predictions than simply using the average score.

# Conclusion

In this notebook, I trained a Linear Regression model to predict students' total scores.

The model was evaluated using MAE, RMSE, and R², and its performance was compared with a baseline model.

The results show that Linear Regression can learn the relationship between the selected features and the target variable and provide meaningful predictions.